Index: u-boot-2020.04/include/configs/imx6_spl.h
===================================================================
--- u-boot-2020.04.orig/include/configs/imx6_spl.h
+++ u-boot-2020.04/include/configs/imx6_spl.h
@@ -67,11 +67,7 @@
 
 /* Define the payload for FAT/EXT support */
 #if defined(CONFIG_SPL_FS_FAT) || defined(CONFIG_SPL_FS_EXT4)
-# ifdef CONFIG_OF_CONTROL
-#  define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME	"u-boot-dtb.img"
-# else
-#  define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME	"u-boot.img"
-# endif
+#define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME	"u-boot.img"
 #endif
 
 #if defined(CONFIG_MX6SX) || defined(CONFIG_MX6SL) || \
