Index: backports-5.4-rc8-1/backport-include/linux/gfp.h
===================================================================
--- backports-5.4-rc8-1.orig/backport-include/linux/gfp.h
+++ backports-5.4-rc8-1/backport-include/linux/gfp.h
@@ -10,7 +10,7 @@
 #define __GFP_KSWAPD_RECLAIM	((__force gfp_t)___GFP_KSWAPD_RECLAIM) /* kswapd can wake */
 #endif
 
-#if LINUX_VERSION_IS_LESS(4,10,0) && LINUX_VERSION_IS_GEQ(4,2,0)
+#if LINUX_VERSION_IS_LESS(4,9,116) && LINUX_VERSION_IS_GEQ(4,2,0)
 #define page_frag_alloc LINUX_BACKPORT(page_frag_alloc)
 static inline void *page_frag_alloc(struct page_frag_cache *nc,
 				    unsigned int fragsz, gfp_t gfp_mask)
