Index: linux-5.10.27/drivers/net/wireless/realtek/rtlwifi/rtl8192cu/sw.c
===================================================================
--- linux-5.10.27.orig/drivers/net/wireless/realtek/rtlwifi/rtl8192cu/sw.c
+++ linux-5.10.27/drivers/net/wireless/realtek/rtlwifi/rtl8192cu/sw.c
@@ -354,6 +354,7 @@ static const struct usb_device_id rtl819
 	{RTL_USB_DEVICE(0x0bda, 0x8186, rtl92cu_hal_cfg)}, /*Realtek 92CE-VAU*/
 	{RTL_USB_DEVICE(0x0df6, 0x0061, rtl92cu_hal_cfg)}, /*Sitecom-Edimax*/
 	{RTL_USB_DEVICE(0x0e66, 0x0019, rtl92cu_hal_cfg)}, /*Hawking-Edimax*/
+	{RTL_USB_DEVICE(0x1058, 0x0631, rtl92cu_hal_cfg)}, /*Western Digital*/
 	{RTL_USB_DEVICE(0x2001, 0x3307, rtl92cu_hal_cfg)}, /*D-Link-Cameo*/
 	{RTL_USB_DEVICE(0x2001, 0x3309, rtl92cu_hal_cfg)}, /*D-Link-Alpha*/
 	{RTL_USB_DEVICE(0x2001, 0x330a, rtl92cu_hal_cfg)}, /*D-Link-Alpha*/
