Index: linux-5.10.83/crypto/zstd.c
===================================================================
--- linux-5.10.83.orig/crypto/zstd.c
+++ linux-5.10.83/crypto/zstd.c
@@ -24,7 +24,7 @@ struct zstd_ctx {
 	void *dwksp;
 };
 
-static ZSTD_parameters zstd_params(void)
+static ZSTD_parameters zstd_params(size_t src_len)
 {
 	return ZSTD_getParams(ZSTD_DEF_LEVEL, 0, 0);
 }
@@ -32,7 +32,7 @@ static ZSTD_parameters zstd_params(void)
 static int zstd_comp_init(struct zstd_ctx *ctx)
 {
 	int ret = 0;
-	const ZSTD_parameters params = zstd_params();
+	const ZSTD_parameters params = zstd_params(0);
 	const size_t wksp_size = ZSTD_CCtxWorkspaceBound(params.cParams);
 
 	ctx->cwksp = vzalloc(wksp_size);
@@ -152,7 +152,7 @@ static int __zstd_compress(const u8 *src
 {
 	size_t out_len;
 	struct zstd_ctx *zctx = ctx;
-	const ZSTD_parameters params = zstd_params();
+	const ZSTD_parameters params = zstd_params(slen);
 
 	out_len = ZSTD_compressCCtx(zctx->cctx, dst, *dlen, src, slen, params);
 	if (ZSTD_isError(out_len))
