Index: netifd-2021-10-21-f78bdec2/device.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.c
+++ netifd-2021-10-21-f78bdec2/device.c
@@ -63,6 +63,7 @@ static const struct blobmsg_policy dev_a
 	[DEV_ATTR_AUTH] = { .name = "auth", .type = BLOBMSG_TYPE_BOOL },
 	[DEV_ATTR_SPEED] = { .name = "speed", .type = BLOBMSG_TYPE_INT32 },
 	[DEV_ATTR_DUPLEX] = { .name = "duplex", .type = BLOBMSG_TYPE_BOOL },
+	[DEV_ATTR_DIS_CHKSUM_OFFLOAD] = { .name = "dis_chksum_offload", .type = BLOBMSG_TYPE_BOOL },
 };
 
 const struct uci_blob_param_list device_attr_list = {
@@ -280,6 +281,8 @@ device_merge_settings(struct device *dev
 	n->auth = s->flags & DEV_OPT_AUTH ? s->auth : os->auth;
 	n->speed = s->flags & DEV_OPT_SPEED ? s->speed : os->speed;
 	n->duplex = s->flags & DEV_OPT_DUPLEX ? s->duplex : os->duplex;
+	n->dis_chksum_offload = s->flags & DEV_OPT_DIS_CHKSUM_OFFLOAD ?
+		s->dis_chksum_offload : os->dis_chksum_offload;
 	n->flags = s->flags | os->flags | os->valid_flags;
 }
 
@@ -464,6 +467,11 @@ device_init_settings(struct device *dev,
 		s->flags |= DEV_OPT_DUPLEX;
 	}
 
+	if ((cur = tb[DEV_ATTR_DIS_CHKSUM_OFFLOAD])) {
+		s->dis_chksum_offload = blobmsg_get_bool(cur);
+		s->flags |= DEV_OPT_DIS_CHKSUM_OFFLOAD;
+	}
+
 	device_set_disabled(dev, disabled);
 }
 
@@ -1210,6 +1218,8 @@ device_dump_status(struct blob_buf *b, s
 			blobmsg_add_u8(b, "arp_accept", st.arp_accept);
 		if (st.flags & DEV_OPT_AUTH)
 			blobmsg_add_u8(b, "auth", st.auth);
+		if (st.flags & DEV_OPT_DIS_CHKSUM_OFFLOAD)
+			blobmsg_add_u8(b, "dis_chksum_offload", st.dis_chksum_offload);
 	}
 
 	s = blobmsg_open_table(b, "statistics");
Index: netifd-2021-10-21-f78bdec2/device.h
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.h
+++ netifd-2021-10-21-f78bdec2/device.h
@@ -62,6 +62,7 @@ enum {
 	DEV_ATTR_AUTH,
 	DEV_ATTR_SPEED,
 	DEV_ATTR_DUPLEX,
+	DEV_ATTR_DIS_CHKSUM_OFFLOAD,
 	__DEV_ATTR_MAX,
 };
 
@@ -126,6 +127,7 @@ enum {
 	DEV_OPT_ARP_ACCEPT		= (1ULL << 29),
 	DEV_OPT_SPEED			= (1ULL << 30),
 	DEV_OPT_DUPLEX			= (1ULL << 31),
+	DEV_OPT_DIS_CHKSUM_OFFLOAD	= (1ULL << 32),
 };
 
 /* events broadcasted to all users of a device */
@@ -203,6 +205,7 @@ struct device_settings {
 	bool auth;
 	unsigned int speed;
 	bool duplex;
+	bool dis_chksum_offload;
 };
 
 /*
Index: netifd-2021-10-21-f78bdec2/system-linux.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/system-linux.c
+++ netifd-2021-10-21-f78bdec2/system-linux.c
@@ -1897,6 +1897,18 @@ system_if_get_settings(struct device *de
 		s->arp_accept = strtoul(buf, NULL, 0);
 		s->flags |= DEV_OPT_ARP_ACCEPT;
 	}
+
+	struct ethtool_value ecmd;
+	memset(&ecmd, 0, sizeof(ecmd));
+	ifr.ifr_data = (caddr_t) &ecmd;
+	ecmd.cmd = ETHTOOL_GRXCSUM;
+	if (!ioctl(sock_ioctl, SIOCETHTOOL, &ifr) && !ecmd.data){
+		ecmd.cmd = ETHTOOL_GTXCSUM;
+		if (!ioctl(sock_ioctl, SIOCETHTOOL, &ifr) && !ecmd.data){
+			s->dis_chksum_offload=1;
+			s->flags |= DEV_OPT_DIS_CHKSUM_OFFLOAD;
+		}
+	}
 }
 
 void
@@ -1995,6 +2007,18 @@ system_if_apply_settings(struct device *
 		system_set_drop_unsolicited_na(dev, s->drop_unsolicited_na ? "1" : "0");
 	if (apply_mask & DEV_OPT_ARP_ACCEPT)
 		system_set_arp_accept(dev, s->arp_accept ? "1" : "0");
+
+	if (s->flags & DEV_OPT_DIS_CHKSUM_OFFLOAD & apply_mask){
+		struct ethtool_value ecmd;
+		memset(&ecmd, 0, sizeof(ecmd));
+		ifr.ifr_data = (caddr_t) &ecmd;
+		ecmd.cmd = ETHTOOL_SRXCSUM;
+		ecmd.data=!s->dis_chksum_offload;
+		ioctl(sock_ioctl, SIOCETHTOOL, &ifr);
+		ecmd.cmd = ETHTOOL_STXCSUM;
+		ioctl(sock_ioctl, SIOCETHTOOL, &ifr);
+	}
+
 	system_set_ethtool_settings(dev, s);
 }
 
@@ -3545,7 +3569,7 @@ static void system_vxlan_map_bool_attr(s
 
 		if ((attrtype == IFLA_VXLAN_GBP) && val)
 			nla_put_flag(msg, attrtype);
-		else 
+		else
 			nla_put_u8(msg, attrtype, val);
 
 	}
Index: netifd-2021-10-21-f78bdec2/system.h
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/system.h
+++ netifd-2021-10-21-f78bdec2/system.h
@@ -273,7 +273,6 @@ int system_if_down(struct device *dev);
 int system_if_check(struct device *dev);
 int system_if_resolve(struct device *dev);
 int system_if_flags(const char *ifname, unsigned add, unsigned rem);
-
 int system_if_dump_info(struct device *dev, struct blob_buf *b);
 int system_if_dump_stats(struct device *dev, struct blob_buf *b);
 struct device *system_if_get_parent(struct device *dev);
