Index: netifd-2021-10-21-f78bdec2/device.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.c
+++ netifd-2021-10-21-f78bdec2/device.c
@@ -64,6 +64,7 @@ static const struct blobmsg_policy dev_a
 	[DEV_ATTR_SPEED] = { .name = "speed", .type = BLOBMSG_TYPE_INT32 },
 	[DEV_ATTR_DUPLEX] = { .name = "duplex", .type = BLOBMSG_TYPE_BOOL },
 	[DEV_ATTR_DIS_CHKSUM_OFFLOAD] = { .name = "dis_chksum_offload", .type = BLOBMSG_TYPE_BOOL },
+	[DEV_ATTR_PROXYARP] = { .name = "proxy_arp", .type = BLOBMSG_TYPE_BOOL },
 };
 
 const struct uci_blob_param_list device_attr_list = {
@@ -244,6 +245,7 @@ device_merge_settings(struct device *dev
 	n->ip6segmentrouting = s->flags & DEV_OPT_IP6SEGMENTROUTING ? s->ip6segmentrouting : os->ip6segmentrouting;
 	n->promisc = s->flags & DEV_OPT_PROMISC ? s->promisc : os->promisc;
 	n->rpfilter = s->flags & DEV_OPT_RPFILTER ? s->rpfilter : os->rpfilter;
+	n->proxy_arp = s->flags & DEV_OPT_PROXYARP? s->proxy_arp : os->proxy_arp;
 	n->acceptlocal = s->flags & DEV_OPT_ACCEPTLOCAL ? s->acceptlocal : os->acceptlocal;
 	n->igmpversion = s->flags & DEV_OPT_IGMPVERSION ? s->igmpversion : os->igmpversion;
 	n->mldversion = s->flags & DEV_OPT_MLDVERSION ? s->mldversion : os->mldversion;
@@ -336,6 +338,11 @@ device_init_settings(struct device *dev,
 		s->flags |= DEV_OPT_PROMISC;
 	}
 
+	if ((cur = tb[DEV_ATTR_PROXYARP])) {
+		s->proxy_arp = blobmsg_get_bool(cur);
+		s->flags |= DEV_OPT_PROXYARP;
+	}
+
 	if ((cur = tb[DEV_ATTR_RPFILTER])) {
 		if (system_resolve_rpfilter(blobmsg_data(cur), &s->rpfilter))
 			s->flags |= DEV_OPT_RPFILTER;
@@ -1220,6 +1227,8 @@ device_dump_status(struct blob_buf *b, s
 			blobmsg_add_u8(b, "auth", st.auth);
 		if (st.flags & DEV_OPT_DIS_CHKSUM_OFFLOAD)
 			blobmsg_add_u8(b, "dis_chksum_offload", st.dis_chksum_offload);
+		if (st.flags & DEV_OPT_PROXYARP)
+			blobmsg_add_u8(b, "proxy_arp", st.proxy_arp);
 	}
 
 	s = blobmsg_open_table(b, "statistics");
Index: netifd-2021-10-21-f78bdec2/device.h
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.h
+++ netifd-2021-10-21-f78bdec2/device.h
@@ -39,6 +39,7 @@ enum {
 	DEV_ATTR_PROMISC,
 	DEV_ATTR_RPFILTER,
 	DEV_ATTR_ACCEPTLOCAL,
+	DEV_ATTR_PROXYARP,
 	DEV_ATTR_IGMPVERSION,
 	DEV_ATTR_MLDVERSION,
 	DEV_ATTR_NEIGHREACHABLETIME,
@@ -128,6 +129,7 @@ enum {
 	DEV_OPT_SPEED			= (1ULL << 30),
 	DEV_OPT_DUPLEX			= (1ULL << 31),
 	DEV_OPT_DIS_CHKSUM_OFFLOAD	= (1ULL << 32),
+	DEV_OPT_PROXYARP		= (1 << 33),
 };
 
 /* events broadcasted to all users of a device */
@@ -206,6 +208,7 @@ struct device_settings {
 	unsigned int speed;
 	bool duplex;
 	bool dis_chksum_offload;
+	bool proxy_arp;
 };
 
 /*
Index: netifd-2021-10-21-f78bdec2/system-linux.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/system-linux.c
+++ netifd-2021-10-21-f78bdec2/system-linux.c
@@ -391,6 +391,11 @@ static void system_set_acceptlocal(struc
 	system_set_dev_sysctl("ipv4/conf", "accept_local", dev->ifname, val);
 }
 
+static void system_set_proxy_arp(struct device *dev, const char *val)
+{
+	system_set_dev_sysctl("/proc/sys/net/ipv4/conf/%s/proxy_arp", dev->ifname, val);
+}
+
 static void system_set_igmpversion(struct device *dev, const char *val)
 {
 	system_set_dev_sysctl("ipv4/conf", "force_igmp_version", dev->ifname, val);
@@ -613,6 +618,12 @@ static int system_get_acceptlocal(struct
 				     dev->ifname, buf, buf_sz);
 }
 
+static int system_get_proxy_arp(struct device *dev, char *buf, const size_t buf_sz)
+{
+	return system_get_dev_sysctl("/proc/sys/net/ipv4/conf/%s/proxy_arp",
+			dev->ifname, buf, buf_sz);
+}
+
 static int system_get_igmpversion(struct device *dev, char *buf, const size_t buf_sz)
 {
 	return system_get_dev_sysctl("ipv4/conf", "force_igmp_version",
@@ -1828,6 +1839,11 @@ system_if_get_settings(struct device *de
 		s->flags |= DEV_OPT_ACCEPTLOCAL;
 	}
 
+	if (!system_get_proxy_arp(dev, buf, sizeof(buf))) {
+		s->proxy_arp = strtoul(buf, NULL, 0);
+		s->flags |= DEV_OPT_PROXYARP;
+	}
+
 	if (!system_get_igmpversion(dev, buf, sizeof(buf))) {
 		s->igmpversion = strtoul(buf, NULL, 0);
 		s->flags |= DEV_OPT_IGMPVERSION;
@@ -1962,6 +1978,10 @@ system_if_apply_settings(struct device *
 	}
 	if (apply_mask & DEV_OPT_ACCEPTLOCAL)
 		system_set_acceptlocal(dev, s->acceptlocal ? "1" : "0");
+
+	if (DEV_OPT_PROXYARP & apply_mask)
+		system_set_proxy_arp(dev, s->proxy_arp ? "1" : "0");
+
 	if (apply_mask & DEV_OPT_IGMPVERSION) {
 		snprintf(buf, sizeof(buf), "%u", s->igmpversion);
 		system_set_igmpversion(dev, buf);
@@ -2008,7 +2028,7 @@ system_if_apply_settings(struct device *
 	if (apply_mask & DEV_OPT_ARP_ACCEPT)
 		system_set_arp_accept(dev, s->arp_accept ? "1" : "0");
 
-	if (s->flags & DEV_OPT_DIS_CHKSUM_OFFLOAD & apply_mask){
+	if (apply_mask & DEV_OPT_DIS_CHKSUM_OFFLOAD){
 		struct ethtool_value ecmd;
 		memset(&ecmd, 0, sizeof(ecmd));
 		ifr.ifr_data = (caddr_t) &ecmd;
