Index: netifd-2021-07-26-440eb064/system-linux.c
===================================================================
--- netifd-2021-07-26-440eb064.orig/system-linux.c
+++ netifd-2021-07-26-440eb064/system-linux.c
@@ -3560,6 +3560,12 @@ static int system_add_vxlan(const char *
 	else
 		return -EINVAL;
 
+	/* TODO- work with list of vnis- convert to string
+	if (!((cur = tb_data[VXLAN_DATA_ATTR_ID])))
+		return -EINVAL;
+
+	 */
+
 	msg = nlmsg_alloc_simple(RTM_NEWLINK, NLM_F_REQUEST | NLM_F_CREATE | NLM_F_EXCL);
 
 	if (!msg)
@@ -3599,6 +3605,7 @@ static int system_add_vxlan(const char *
 	if (link)
 		nla_put_u32(msg, IFLA_VXLAN_LINK, link);
 
+	/* TODO convert to list of vnis  */
 	if ((cur = tb_data[VXLAN_DATA_ATTR_ID])) {
 		uint32_t id = blobmsg_get_u32(cur);
 		if (id >= (1u << 24) - 1) {
@@ -3656,7 +3663,17 @@ static int system_add_vxlan(const char *
 	}
 	nla_put_u16(msg, IFLA_VXLAN_PORT, htons(port));
 
-	if ((cur = tb_data[VXLAN_DATA_ATTR_SRCPORTMIN])) {
+	if ((cur = tb_data[VXLAN_DATA_ATTR_SPORT])) {
+		uint32_t low,high,vals;
+		if((vals=sscanf(blobmsg_get_string(cur), "%u %u", &low,&high))){
+			if(vals==1)high=low;
+			struct ifla_vxlan_port_range ports = {
+				.low =  htons(low),
+				.high = htons(high),
+			};
+			nla_put(msg, IFLA_VXLAN_PORT_RANGE, sizeof(ports), &ports);
+		}
+	} else if ((cur = tb_data[VXLAN_DATA_ATTR_SRCPORTMIN])) {
 		struct ifla_vxlan_port_range srcports = {0,0};
 
 		uint32_t low = blobmsg_get_u32(cur);
@@ -3969,6 +3986,8 @@ int system_add_ip_tunnel(const char *nam
 #ifdef IFLA_VXLAN_MAX
 	} else if(!strcmp(str, "vxlan")) {
 		return system_add_vxlan(name, link, tb, false);
+	} else if(!strcmp(str, "lanpipe")) {
+		return system_add_vxlan(name, link, tb, false);
 	} else if(!strcmp(str, "vxlan6")) {
 		return system_add_vxlan(name, link, tb, true);
 #endif
Index: netifd-2021-07-26-440eb064/system.c
===================================================================
--- netifd-2021-07-26-440eb064.orig/system.c
+++ netifd-2021-07-26-440eb064/system.c
@@ -34,6 +34,7 @@ const struct uci_blob_param_list tunnel_
 
 static const struct blobmsg_policy vxlan_data_attrs[__VXLAN_DATA_ATTR_MAX] = {
 	[VXLAN_DATA_ATTR_ID] = { .name = "id", .type = BLOBMSG_TYPE_INT32 },
+	[VXLAN_DATA_ATTR_SPORT] = { .name = "sport", .type = BLOBMSG_TYPE_STRING },
 	[VXLAN_DATA_ATTR_PORT] = { .name = "port", .type = BLOBMSG_TYPE_INT32 },
 	[VXLAN_DATA_ATTR_MACADDR] = { .name = "macaddr", .type = BLOBMSG_TYPE_STRING },
 	[VXLAN_DATA_ATTR_RXCSUM] = { .name = "rxcsum", .type = BLOBMSG_TYPE_BOOL },
Index: netifd-2021-07-26-440eb064/system.h
===================================================================
--- netifd-2021-07-26-440eb064.orig/system.h
+++ netifd-2021-07-26-440eb064/system.h
@@ -40,6 +40,7 @@ extern const struct uci_blob_param_list
 
 enum vxlan_data {
 	VXLAN_DATA_ATTR_ID,
+	VXLAN_DATA_ATTR_SPORT,
 	VXLAN_DATA_ATTR_PORT,
 	VXLAN_DATA_ATTR_MACADDR,
 	VXLAN_DATA_ATTR_RXCSUM,
