Index: netifd-2021-10-21-f78bdec2/interface.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/interface.c
+++ netifd-2021-10-21-f78bdec2/interface.c
@@ -59,6 +59,7 @@ enum {
 	IFACE_ATTR_DELEGATE,
 	IFACE_ATTR_IP6IFACEID,
 	IFACE_ATTR_FORCE_LINK,
+	IFACE_ATTR_EXT_LINK,
 	IFACE_ATTR_IP6WEIGHT,
 	IFACE_ATTR_MAX
 };
@@ -89,6 +90,7 @@ static const struct blobmsg_policy iface
 	[IFACE_ATTR_DELEGATE] = { .name = "delegate", .type = BLOBMSG_TYPE_BOOL },
 	[IFACE_ATTR_IP6IFACEID] = { .name = "ip6ifaceid", .type = BLOBMSG_TYPE_STRING },
 	[IFACE_ATTR_FORCE_LINK] = { .name = "force_link", .type = BLOBMSG_TYPE_BOOL },
+	[IFACE_ATTR_EXT_LINK] = { .name = "ext_link", .type = BLOBMSG_TYPE_BOOL },
 	[IFACE_ATTR_IP6WEIGHT] = { .name = "ip6weight", .type = BLOBMSG_TYPE_INT32 },
 };
 
@@ -898,6 +900,7 @@ interface_alloc(const char *name, struct
 
 	iface->autostart = blobmsg_get_bool_default(tb[IFACE_ATTR_AUTO], true);
 	iface->force_link = blobmsg_get_bool_default(tb[IFACE_ATTR_FORCE_LINK], force_link);
+	iface->ext_link = blobmsg_get_bool_default(tb[IFACE_ATTR_EXT_LINK], false);
 	iface->dynamic = dynamic;
 	iface->proto_ip.no_defaultroute =
 		!blobmsg_get_bool_default(tb[IFACE_ATTR_DEFAULTROUTE], true);
@@ -1085,6 +1088,9 @@ interface_set_main_dev(struct interface
 {
 	bool claimed = iface->l3_dev.claimed;
 
+	if (dev && iface->ext_link)
+		dev->ext_link = true;
+
 	if (iface->main_dev.dev == dev)
 		return;
 
Index: netifd-2021-10-21-f78bdec2/interface.h
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/interface.h
+++ netifd-2021-10-21-f78bdec2/interface.h
@@ -128,6 +128,7 @@ struct interface {
 	bool enabled;
 	bool link_state;
 	bool force_link;
+	bool ext_link;
 	bool dynamic;
 	bool policy;
 	bool link_up_event;
Index: netifd-2021-10-21-f78bdec2/device.c
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.c
+++ netifd-2021-10-21-f78bdec2/device.c
@@ -157,7 +157,7 @@ static int set_device_state(struct devic
 		system_if_apply_settings(dev, &dev->settings, dev->settings.flags);
 
 		system_if_up(dev);
-	} else {
+	} else if(!dev->ext_link){
 		system_if_down(dev);
 		system_if_apply_settings(dev, &dev->orig_settings, dev->orig_settings.flags);
 	}
Index: netifd-2021-10-21-f78bdec2/device.h
===================================================================
--- netifd-2021-10-21-f78bdec2.orig/device.h
+++ netifd-2021-10-21-f78bdec2/device.h
@@ -243,6 +243,7 @@ struct device {
 	bool disabled;
 	bool deferred;
 	bool hidden;
+	bool ext_link; /* don't take link down here */
 
 	bool current_config;
 	bool iface_config;
