Index: linux-4.19.82/arch/arm/boot/dts/bcm2708-rpi-zero-w.dts
===================================================================
--- linux-4.19.82.orig/arch/arm/boot/dts/bcm2708-rpi-zero-w.dts
+++ linux-4.19.82/arch/arm/boot/dts/bcm2708-rpi-zero-w.dts
@@ -87,6 +87,16 @@
 	pinctrl-0 = <&uart0_pins &bt_pins>;
 	status = "okay";
 };
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pins &bt_pins>;
+	status = "okay";
+
+	bluetooth {
+		compatible = "brcm,bcm43438-bt";
+		max-speed = <2000000>;
+	};
+};
 
 &uart1 {
 	pinctrl-names = "default";
