Index: linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
===================================================================
--- linux-5.10.27.orig/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
+++ linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
@@ -23,7 +23,7 @@
 	leds {
 		compatible = "gpio-leds";
 		pinctrl-names = "default";
-		pinctrl-0 = <&green_led_gio>, <&heartbeat_led_gpio>;
+		pinctrl-0 = <&green_led_gpio>, <&heartbeat_led_gpio>;
 
 		green-led {
 			label = "rockpis:green:power";
@@ -35,11 +35,32 @@
 		blue-led {
 			label = "rockpis:blue:user";
 			gpios = <&gpio0 RK_PA5 GPIO_ACTIVE_HIGH>;
-			default-state = "on";
-			linux,default-trigger = "heartbeat";
+			linux,default-trigger = "ide-disk";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_keys>;
+
+		volume-button {
+			debounce-interval = <20>;
+			gpios = <&gpio2 RK_PA7 GPIO_ACTIVE_LOW>;
+			label = "Volume Button";
+			linux,code = <119>; /* KEY_PAUSE */
 		};
 	};
 
+	volume {
+		compatible = "rotary-encoder";
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_rotary>;
+		gpios = <&gpio1 RK_PC7 GPIO_ACTIVE_LOW>, <&gpio1 RK_PC6 GPIO_ACTIVE_LOW>;
+		linux,axis = <1>; /* REL_Y */
+		rotary-encoder,relative-axis;
+	};
+
 	vcc5v0_sys: vcc5v0-sys {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc5v0_sys";
@@ -93,7 +114,6 @@
 		vin-supply = <&vcc5v0_sys>;
 	};
 
-
 	vcc_1v8: vcc-1v8 {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc_1v8";
@@ -195,7 +215,7 @@
 	pinctrl-0 = <&rtc_32k>;
 
 	leds {
-		green_led_gio: green-led-gpio {
+		green_led_gpio: green-led-gpio {
 			rockchip,pins = <0 RK_PA6 RK_FUNC_GPIO &pcfg_pull_none>;
 		};
 
@@ -209,6 +229,19 @@
 			rockchip,pins = <0 RK_PC5 RK_FUNC_GPIO &pcfg_pull_none>;
 		};
 	};
+
+	keys {
+		gpio_keys: gpio-key-volume-button {
+			rockchip,pins = <2 RK_PA7 RK_FUNC_GPIO &pcfg_pull_up_4ma>;
+		};
+	};
+
+	volume {
+		gpio_rotary: gpio-volume {
+			rockchip,pins = <1 RK_PC7 RK_FUNC_GPIO &pcfg_pull_up_4ma>,
+					<1 RK_PC6 RK_FUNC_GPIO &pcfg_pull_up_4ma>;
+		};
+	};
 };
 
 &pwm0 {
@@ -234,6 +267,10 @@
 	status = "okay";
 };
 
+&spdif_out {
+	status = "okay";
+};
+
 &spi2 {
 	status = "okay";
 	max-freq = <10000000>;
