Index: linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
===================================================================
--- linux-5.10.27.orig/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
+++ linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
@@ -69,6 +92,23 @@
 		};
 	};
 
+	spdif: spdif {
+		compatible = "linux,spdif-dit";
+		#sound-dai-cells = <0>;
+	};
+
+	spdif_sound: spdif-sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "SPDIF";
+		simple-audio-card,cpu {
+			sound-dai = <&spdif_tx>;
+		};
+
+		simple-audio-card,codec {
+			sound-dai = <&spdif>;
+		};
+	};
+
 	pcm5102a: pcm5102a {
 			#sound-dai-cells = <0>;
 			compatible = "ti,pcm5102a";
@@ -352,3 +396,8 @@
 &usb_host_ohci{
 	status = "okay";
 };
+
+&spdif_tx {
+    status = "okay";
+    #sound-dai-cells = <0>;
+};
