Index: linux-5.10.27/sound/soc/codecs/rk3308_codec.c
===================================================================
--- linux-5.10.27.orig/sound/soc/codecs/rk3308_codec.c
+++ linux-5.10.27/sound/soc/codecs/rk3308_codec.c
@@ -244,8 +244,6 @@ static const DECLARE_TLV_DB_SCALE(rk3308
 				  -600, 150, 0);
 static const DECLARE_TLV_DB_SCALE(rk3308_codec_dac_hpout_gain_tlv,
 				  -3900, 150, 600);
-static const DECLARE_TLV_DB_SCALE(rk3308_codec_dac_hpmix_gain_tlv,
-				  -600, 600, 0);
 
 static const DECLARE_TLV_DB_RANGE(rk3308_codec_adc_mic_gain_tlv_a,
 	0, 0, TLV_DB_SCALE_ITEM(0, 0, 0),
@@ -864,20 +862,6 @@ static const struct snd_kcontrol_new rk3
 			   rk3308_codec_hpout_r_get_tlv,
 			   rk3308_codec_hpout_r_put_tlv,
 			   rk3308_codec_dac_hpout_gain_tlv),
-
-	/* DAC HPMIX */
-	SOC_SINGLE_RANGE_TLV("DAC HPMIX Left Volume",
-			     RK3308_DAC_ANA_CON12,
-			     RK3308_DAC_L_HPMIX_GAIN_SFT,
-			     RK3308_DAC_L_HPMIX_GAIN_MIN,
-			     RK3308_DAC_L_HPMIX_GAIN_MAX,
-			     0, rk3308_codec_dac_hpmix_gain_tlv),
-	SOC_SINGLE_RANGE_TLV("DAC HPMIX Right Volume",
-			     RK3308_DAC_ANA_CON12,
-			     RK3308_DAC_R_HPMIX_GAIN_SFT,
-			     RK3308_DAC_R_HPMIX_GAIN_MIN,
-			     RK3308_DAC_R_HPMIX_GAIN_MAX,
-			     0, rk3308_codec_dac_hpmix_gain_tlv),
 };
 
 static int rk3308_codec_agc_get(struct snd_kcontrol *kcontrol,
@@ -1728,7 +1712,7 @@ static int rk3308_mute_stream(struct snd
 		int dgain;
 
 		if (mute) {
-			for (dgain = 0x2; dgain <= 0x7; dgain++) {
+			for (dgain = 0x1; dgain <= 0x7; dgain++) {
 				/*
 				 * Keep the max -> min digital CIC interpolation
 				 * filter gain step by step.
@@ -1756,7 +1740,7 @@ static int rk3308_mute_stream(struct snd
 			if (rk3308->delay_start_play_ms)
 				msleep(rk3308->delay_start_play_ms);
 #endif
-			for (dgain = 0x7; dgain >= 0x2; dgain--) {
+			for (dgain = 0x7; dgain >= 0x1; dgain--) {
 				/*
 				 * Keep the min -> max digital CIC interpolation
 				 * filter gain step by step
@@ -1855,6 +1839,10 @@ static int rk3308_codec_digital_fadeout(
 
 static int rk3308_codec_dac_lineout_enable(struct rk3308_codec_priv *rk3308)
 {
+	/* Increase Line Out strength */
+	regmap_update_bits(rk3308->regmap, RK3308_DAC_ANA_CON08,
+			   RK3308_DAC_R_LINEOUT_DRV_MSK|RK3308_DAC_L_LINEOUT_DRV_MSK, 0x11);
+
 	if (rk3308->codec_ver == ACODEC_VERSION_B) {
 		/* Step 04 */
 		regmap_update_bits(rk3308->regmap, RK3308_DAC_ANA_CON15,
@@ -2128,8 +2116,8 @@ static int rk3308_codec_dac_enable(struc
 	regmap_update_bits(rk3308->regmap, RK3308_DAC_ANA_CON13,
 			   RK3308_DAC_L_HPMIX_EN |
 			   RK3308_DAC_R_HPMIX_EN,
-			   RK3308_DAC_L_HPMIX_EN |
-			   RK3308_DAC_R_HPMIX_EN);
+			   RK3308_DAC_L_HPMIX_DIS |
+			   RK3308_DAC_R_HPMIX_DIS);
 
 	/* Waiting the stable HPMIX */
 	mdelay(1);
@@ -3908,8 +3896,8 @@ static int rk3308_codec_default_gains(st
 	regmap_update_bits(rk3308->regmap, RK3308_DAC_ANA_CON04,
 			   RK3308_DAC_L_LINEOUT_GAIN_MSK |
 			   RK3308_DAC_R_LINEOUT_GAIN_MSK,
-			   RK3308_DAC_L_LINEOUT_GAIN_NDB_6 |
-			   RK3308_DAC_R_LINEOUT_GAIN_NDB_6);
+			   RK3308_DAC_L_LINEOUT_GAIN_0DB |
+			   RK3308_DAC_R_LINEOUT_GAIN_0DB);
 
 	return 0;
 }
Index: linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
===================================================================
--- linux-5.10.27.orig/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
+++ linux-5.10.27/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
@@ -69,6 +92,23 @@
 		};
 	};
 
+	spdif: spdif {
+		compatible = "linux,spdif-dit";
+		#sound-dai-cells = <0>;
+	};
+
+	spdif_sound: spdif-sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "SPDIF";
+		simple-audio-card,cpu {
+			sound-dai = <&spdif_tx>;
+		};
+
+		simple-audio-card,codec {
+			sound-dai = <&spdif>;
+		};
+	};
+
 	pcm5102a: pcm5102a {
 			#sound-dai-cells = <0>;
 			compatible = "ti,pcm5102a";
@@ -352,3 +396,8 @@
 &usb_host_ohci{
 	status = "okay";
 };
+
+&spdif_tx {
+    status = "okay";
+    #sound-dai-cells = <0>;
+};
