From 19e34504d0067dd5fcaa2323eb13e8cb7401a0ad Mon Sep 17 00:00:00 2001
From: Clark Williams <williams@redhat.com>
Date: Sat, 25 Apr 2020 20:29:17 -0500
Subject: [PATCH 366/366] Linux 4.9.220-rt143 REBASE

Signed-off-by: Clark Williams <williams@redhat.com>
---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index 9a218ca23053..a7cc5d609b19 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt23
+-rt143
-- 
2.25.3

